apps/src/gamelab/PropTypes.js

```diff
@@ -0,0 +1,149 @@
+/** @file Type definitions (React and otherwise) specific to Gamelab */
+
+import _ from 'lodash';
+import React from 'react';
+
+/**
+ * @typedef {Object} Vector2
+ * @property {number} x
+ * @property {number} y
+ */
+const Vector2 = React.PropTypes.shape({
+  x: React.PropTypes.number.isRequired,
+  y: React.PropTypes.number.isRequired
+});
+
+/*
+ * MIGRATION NOTES - 30 June 2016 bbuchanan
+ *  The old shape of animations is an array of animation metadata.
+ *    [
+ *      {animation},
+ *      {animation}
+ *    ]
+ *
+ *  The new shape is an object with animation and cache components
+ *    {
+ *     orderedKeys: [ AnimationKey, AnimationKey ],
+ *     propsByKey: {
+ *       AnimationKey: {AnimationProps},
+ *       AnimationKey: {AnimationProps}
+ *     }
+ *   }
+ *
+ *  AnimationProps should include the actual spritesheet (as blob and dataURI),
+ *  dimensions and frame dimensions, framerate, name, last save time, version IDs,
+ *  URL to fetch the animation from the API, etc.
+ *
+ *  We serialize a smaller set of information {SerializedAnimationProps}.
+ *
+ *  We need to do a migration if the old style gets loaded.
+ *  See setInitialAnimationList for how this works.
+ */
+
+/**
+ * @typedef {string} AnimationKey
+ * A string that uniquely identifies an animation within the project, usually
+ * a UUID.
+ */
+export const AnimationKey = React.PropTypes.string;
+
+/**
+ * @typedef {Object} AnimationProps
+ * @property {string} name
+ * @property {string} sourceUrl - Remote location where the animation spritesheet
+ *           is stored.  Should be replaced soon with more of a 'gallery' concept.
+ * @property {Vector2} sourceSize
+ * @property {Vector2} frameSize
+ * @property {number} frameCount
+ * @property {number} frameRate
+ * @property {string} [version] - S3 version key
+ *
+ * @property {boolean} loadedFromSource - False at first, true after load successful.
+ * @property {boolean} saved - True if the current blob represents the last thing
+ *           we uploaded to the animations API, false if we have a pending change
+ *           to save.
+ * @property {Blob} blob - The spritesheet as a data blob (can be uploaded
+ *           directly to the animations API via PUT).
+ * @property {string} dataURI - The spritesheet as a dataURI (can be set as src
+ *           on an image).
+ * @property {boolean} hasNewVersionThisSession - Whether a new version of the
+ *           animation has been created in the current session. If true, we
+ *           should continue replacing this version instead of writing new
```

Comment:
That's probably because this feature is not implemented yet, and this particular property isn't used anywhere.  I can tear it out until then.

In a versioned S3 bucket you can write to an object (which creates a new version) -or- you can write to a particular version of an object (which replaces that version).  Eventually the goal is for gamelab to do the former (create a new version) only once per animation per session, which is what our project JSON does.  This should minimize the number of versions we're storing on S3.  Right now, we're just writing to the object (and creating a new version) on every save.

This is pretty high on the pre-release priority list, but it's not blocking to the curriculum writers so I'm trying to get this big change out so we can do that small one in its own PR later.